AK+LibJS+LibRegex: Allow Unicode escape sequences in identifiers and regex capture groups

trflynn89 opened the pull request:
The purpose is to allow Unicode escape sequences in both JS identifiers and regex capture group names. For example, the following lines are equivalent:
```javascript
"brown".match(/(?<𝓑𝓻𝓸𝔀𝓷>brown)/u).groups.𝓑𝓻𝓸𝔀𝓷
"brown".match(/(?<\u{1d4d1}\u{1d4fb}\u{1d4f8}\u{1d500}\u{1d4f7}>brown)/u).groups.𝓑𝓻𝓸𝔀𝓷
"brown".match(/(?<𝓑𝓻𝓸𝔀𝓷>brown)/u).groups.\u{1d4d1}\u{1d4fb}\u{1d4f8}\u{1d500}\u{1d4f7}
"brown".match(/(?<\u{1d4d1}\u{1d4fb}\u{1d4f8}\u{1d500}\u{1d4f7}>brown)/u).groups.\u{1d4d1}\u{1d4fb}\u{1d4f8}\u{1d500}\u{1d4f7}
```

Before implementing that, this PR first moves parsing of Unicode escape sequences to `GenericLexer`. This was already implemented in both LibJS ad LibRegex in a few functions, so it was about time to generalize it.

trflynn89 commented:
Marking as a draft for now, there are some test262 regressions with this PR as-is.

The main thing seems to be tests that expect parsing to fail. For example, we used to fail to parse `f\u0061lse` because we weren't parsing the Unicode escape, so the tests that use this variant of `false` in an invalid context would actually pass. We now correctly parse this as `false`, but are not recognizing it as a `BoolLiteral` because resolving that escape happens too late.

This should be easy to fix, but might involve `Token` holding a new `String` object for resolved identifier values.

alimpfard commented:
Make that Token field a `Variant<String, StringView>` (or an equivalent with niceties) to avoid allocating strings unless required - see `Reader::XSV::Field` in Spreadsheet for exactly what I mean).

alimpfard commented on Userland/Libraries/LibRegex/RegexParser.h:
```diff
@@ -53,6 +53,7 @@ class Parser {
         size_t match_length_minimum;
         Error error;
         Token error_token;
+        Vector<FlyString> capture_groups;
```
Since we're now putting these in FlyStrings, let's make use of that as an optimisation and make the match state keys FlyStrings too

davidot commented on Userland/Libraries/LibJS/Lexer.cpp:
```diff
@@ -369,8 +370,22 @@ bool Lexer::is_whitespace() const
+{
+    identifier_length = 1;
+
+    if (m_current_char == '\\')
+        return is_unicode_escape(identifier_length);
```
So here we would need to verify that it would pass the checks below `is_ascii_alpha(...) || == '_'`, etc

trflynn89 replied:
Do we? At this point we know the next six chars are `\unnnn`, where `n` is a hex digit, so also doing the check below seems like extra work.

trflynn89 replied:
Oh, or are you saying the resolved code point needs to pass that test? I see

davidot replied:
Yeah so you cannot get an identifier starting with `0` for example `var \u{30};`

davidot commented on Userland/Libraries/LibJS/Lexer.cpp:
```diff
@@ -381,10 +396,14 @@ bool Lexer::is_identifier_start() const
 {
+    identifier_length = 1;
+
+    if (m_current_char == '\\')
+        return is_unicode_escape(identifier_length);
```
And here check the escaped code point is a valid identifier middle.

davidot commented on Userland/Libraries/LibJS/Lexer.cpp:
```diff
@@ -381,10 +396,14 @@ bool Lexer::is_identifier_start() const
+        return is_unicode_escape(identifier_length);
     if (!is_unicode_character())
-        return is_identifier_start() || is_ascii_digit(m_current_char);
+        return is_identifier_start(identifier_length) || is_ascii_digit(m_current_char);
```
Could this be a problem with the call to `is_identifier_start`? 
Since that could allow characters which are in `ID_Start` but not in `ID_Continue`

davidot replied:
Actually it seems ID_Continue is defined to be a superset of `ID_Start` so this should not be a problem.

davidot commented on Userland/Libraries/LibJS/Parser.cpp:
```diff
@@ -1412,11 +1412,30 @@ NonnullRefPtr<Expression> Parser::parse_secondary_expression(NonnullRefPtr<Expre
             expected("IdentifierName");
-        return create_ast_node<MemberExpression>({ m_state.current_token.filename(), rule_start.position(), position() }, move(lhs), create_ast_node<Identifier>({ m_state.current_token.filename(), rule_start.position(), position() }, consume().value()));
+        } else {
+            auto token = consume();
```
Is this something that every identifier consuming place should do?
In that case we probably want to put it in some method since with this escaped identifiers will not work in other places.

trflynn89 replied:
Yeah actually, I have that change made locally already - I will be storing the resolved identifier value in `Token`, which will return that value in `Token::value`, so all call sets get the right string.

davidot reviewed (changes requested):
Nice work! :smile: 

> The main thing seems to be tests that expect parsing to fail. For example, we used to fail to parse `f\u0061lse` because we weren't parsing the Unicode escape, so the tests that use this variant of `false` in an invalid context would actually pass. We now correctly parse this as `false`, but are not recognizing it as a `BoolLiteral` because resolving that escape happens too late.

Actually I think you're supposed to fail parsing at that point, see Note from https://tc39.es/ecma262/#sec-identifiers-static-semantics-early-errors and https://tc39.es/ecma262/#prod-ReservedWord.

Furthermore we need to do additional verification of the escaped code point since it must be a valid identifier (start) character still see below.

Also if you don't want to check all the identifier consuming places that's fine we can just add member expression for now (there are a lot and they are all slightly different but that is another yak)

trflynn89 commented:
> Actually I think you're supposed to fail parsing at that point, see Note from https://tc39.es/ecma262/#sec-identifiers-static-semantics-early-errors and https://tc39.es/ecma262/#prod-ReservedWord.

Exactly, that's what I meant by passing the test. We "passed" before because the parser failed to handle the Unicode escape. Right now this change is failing the test because it is parsing the escape, but not recognizing the result as a BoolLiteral, so we don't fail parsing the file.

trflynn89 commented:
Alrighty this should be good to go - expected to fix about 300 test262 tests

davidot commented:
All the changes LGTM.

However there are still some things we which don't do correctly.
- We allow identifiers with escapes, now maybe I misunderstood your comment or mine wasn't clear but what I meant was that:
```js
\u{69}\u{66}(true)  // escapes for if(true)
    console.log('should have thrown syntax error');
```
 should not be allowed. 
For example V8 throws: `SyntaxError: Keyword must not contain escaped characters`

- The second issue is that escaping characters which are allowed in the middle of an identifier but not the start.
Because we only escape in `is_identifier_start` as example we fail on
```js
var a\u{30} = 3; // escapes for var a0 = 3;
```

It's also fine to keep the PR as is and we'll just fix these issues in a later PR. Don't want to force extra work on you or anything like that.

trflynn89 commented:
>  We allow identifiers with escapes

Right, and this PR handles that in some cases, but I'm sure it missed some. I'll look around for more cases we can easily catch.

> The second issue is that escaping characters which are allowed in the middle of an identifier but not the start.

Good catch! I must've deleted the escape parsing from `is_identifier_middle` by mistake.

davidot commented:
> > We allow identifiers with escapes
> 
> Right, and this PR handles that in some cases, but I'm sure it missed some. I'll look around for more cases we can easily catch.

Sorry I meant we allow keywords with escapes. Which should not be allowed per the spec. (See the first example in my previous comment)

Again we can also just leave this for a next PR, I currently have one with different parser fixes locally anyway.

trflynn89 commented:
> Sorry I meant we allow keywords with escapes. Which should not be allowed per the spec. (See the first example in my previous comment)

Oh, is there a list of keywords that are never allowed to be escaped? Unfortunately, it depends on context in some cases, because `let yield=0`; / `let \u{79}ield = 0;` are both valid when outside of a generator context.

trflynn89 commented:
Ah, found the text I was missing in the spec earlier:
> All tokens in the ReservedWord list below, except for await and yield, are unconditionally reserved.

I'll just throw that into the last commit, it's an easy thing to just do now. Thanks :)

davidot reviewed (approved):
> > Sorry I meant we allow keywords with escapes. Which should not be allowed per the spec. (See the first example in my previous comment)
> 
> Oh, is there a list of keywords that are never allowed to be escaped? Unfortunately, it depends on context in some cases, because `let yield=0`; / `let \u{79}ield = 0;` are both valid when outside of a generator context.

Oh right that is a problem. Didn't think of that, unfortunately I think most keywords are allowed to be identifiers in some locations. 
~For example V8 gives:~
```js
var a = {};
a.break = 3;
a.function = 3;
a.if = 3;
a.true = 3;
console.log(a);
```
~Gives: `{ break: 3, function: 3, if: 3, true: 3 }`~

Nevermind you're still not allow to escape them

Then let's just ship it as is. Sorry I completely forgot about that.

trflynn89 commented:
Updated the last commit to disallow Unicode escapes in unconditionally reserved keywords. Added bonus that we pass this whole section of tests now:
```
test/language/identifiers    188/188   (100.00%) [ ✅ 188 ]
```
(we were at 149 / 188 before)

trflynn89 commented:
Ah man, well after all that I reverted that change - turns out, the following is valid:
```js
var obj = {
  c\u0061se() { return 42; }
};
```

Tested by `test/language/expressions/object/ident-name-method-def-case-escaped.js` (there's about 35 similar tests in that folder). So let's just handle this in a different PR - we'll have to handle it in the parser somewhere since it's context dependent.
